runtime/src/consensus/registry.rs

```diff
@@ -4,27 +4,243 @@
 };
-use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize_repr, Deserialize_repr)]
+#[repr(u32)]
+pub enum RuntimeKind {
+    #[serde(rename = "invalid")]
+    KindInvalid = 0,
+    #[serde(rename = "compute")]
+    KindCompute = 1,
+    #[serde(rename = "keymanager")]
+    KindKeyManager = 2,
+}
+
+impl Default for RuntimeKind {
+    fn default() -> Self {
+        RuntimeKind::KindInvalid
+    }
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct ExecutorParameters {
+    #[serde(default)]
+    pub group_size: u64,
+    #[serde(default)]
+    pub group_backup_size: u64,
+    #[serde(default)]
+    pub allowed_stragglers: u64,
+    #[serde(default)]
+    pub round_timeout: i64,
+    #[serde(default)]
+    pub max_messages: u32,
+    #[serde(default)]
+    pub min_pool_size: u64,
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct TxnSchedulerParameters {
+    #[serde(default)]
+    pub algorithm: String,
+    #[serde(default)]
+    pub batch_flush_timeout: i64, // In nanoseconds.
+    #[serde(default)]
+    pub max_batch_size: u64,
+    #[serde(default)]
+    pub max_batch_size_bytes: u64,
+    #[serde(default)]
+    pub propose_batch_timeout: i64,
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct StorageParameters {
+    #[serde(default)]
+    pub group_size: u64,
+    #[serde(default)]
+    pub min_write_replication: u64,
+    #[serde(default)]
+    pub max_apply_write_log_entries: u64,
+    #[serde(default)]
+    pub max_apply_ops: u64,
+    #[serde(default)]
+    pub checkpoint_interval: u64,
+    #[serde(default)]
+    pub checkpoint_num_kept: u64,
+    #[serde(default)]
+    pub checkpoint_chunk_size: u64,
+    #[serde(default)]
+    pub min_pool_size: u64,
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct RuntimeStakingParameters {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(default)]
+    pub thresholds: Option<BTreeMap<staking::ThresholdKind, quantity::Quantity>>,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize_repr, Deserialize_repr)]
+#[repr(u32)]
+pub enum RolesMask {
+    #[serde(rename = "compute")]
+    RoleComputeWorker = 1 << 0,
+    #[serde(rename = "storage")]
+    RoleStorageWorker = 1 << 1,
+    #[serde(rename = "key-manager")]
+    RoleKeyManager = 1 << 2,
+    #[serde(rename = "validator")]
+    RoleValidator = 1 << 3,
+    #[serde(rename = "consensus-rpc")]
+    RoleConsensusRPC = 1 << 4,
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct EntityWhitelistRuntimeAdmissionPolicy {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(rename = "entities")]
+    #[serde(default)]
+    pub entities: Option<BTreeMap<PublicKey, EntityWhitelistConfig>>,
```

Comment:
I think you could avoid the `Option` wrapper here and instead do:
```
#[serde(skip_serializing_if = "BTreeMap::is_empty")]
pub entities: BTreeMap<PublicKey, EntityWhitelistConfig>
```
and that should match the `go` serialization. Same for `max_nodes` bellow.

Reply:
I think I tried something similar in the beginning, but it didn't work in all cases -- I'll give it another try to be sure.